akvo/rsr/usecases/run_iati_validation_jobs.py

```diff
@@ -0,0 +1,238 @@
+from django.utils.timezone import now
+
+from akvo.iati.iati_validator import IATIValidatorAPI, IATIValidationResult, IATIValidatorException, IATIValidatorTimeoutException
+from akvo.iati.checks.iati_checks import IatiChecks
+from akvo.iati.exports.iati_export import IatiXML
+from akvo.iati.exports.iati_org_export import IatiOrgXML
+from akvo.rsr.models import Project, Organisation, IatiActivityValidationJob, IatiOrganisationValidationJob, IatiCheck
+from akvo.utils import rsr_send_mail
+
+logger = logging.getLogger(__name__)
+
+VALIDATOR_TIMEOUT = 30  # seconds
+VALIDATOR_MAX_ATTEMPTS = getattr(settings, 'IATI_VALIDATOR_MAX_ATTEMPTS', 3)
+validator = IATIValidatorAPI(getattr(settings, 'IATI_VALIDATOR_SUBSCRIPTION_KEY', ''), VALIDATOR_TIMEOUT)
+
+
+@dataclass(frozen=True)
+class RequestRate:
+    count: int
+    period: timedelta
+
+
+class IATIValidatorRateLimiter:
+
+    def __init__(self, short_term: RequestRate, long_term: RequestRate):
+        self.short_term = short_term
+        self.long_term = long_term
+
+    def is_allowed(self) -> bool:
+        if not self._apply_rate(self.long_term):
+            return False
+        if not self._apply_rate(self.short_term, regulate=True):
+            return False
+        return True
+
+    def _apply_rate(self, rate: RequestRate, regulate: bool = False) -> bool:
+        if rate.count == 0 or rate.period.total_seconds() == 0:
+            return True
+        finished_at = now() - rate.period
+        count, latest = self._get_finished_jobs_summary(finished_at)
+        if count >= rate.count:
+            return False
+        if not regulate:
+            return True
+        # Spread the request evenly in the given time period
+        spacing = timedelta(seconds=rate.period.total_seconds() / rate.count)
+        if latest and latest >= now() - spacing:
+            return False
+        return True
+
+    def _get_finished_jobs_summary(self, finished_at: datetime) -> Tuple[int, Optional[datetime]]:
+        activity_jobs = IatiActivityValidationJob.objects.filter(finished_at__gt=finished_at)
+        organisation_jobs = IatiOrganisationValidationJob.objects.filter(finished_at__gt=finished_at)
+        job_count = activity_jobs.count() + organisation_jobs.count()
+        datetimes = [j.finished_at for j in [
+            activity_jobs.order_by('-finished_at').first(),
+            organisation_jobs.order_by('-finished_at').first()
+        ] if j]
+        latest = max(datetimes) if datetimes else None
+        return (job_count, latest)
+
+
+rate_limiter = IATIValidatorRateLimiter(
+    short_term=RequestRate(
+        count=getattr(settings, 'IATI_VALIDATOR_RATE_LIMIT_COUNT', 0),
+        period=getattr(settings, 'IATI_VALIDATOR_RATE_LIMIT_PERIOD', timedelta(seconds=0))
+    ),
+    long_term=RequestRate(
+        count=getattr(settings, 'IATI_VALIDATOR_QUOTA_COUNT', 0),
+        period=getattr(settings, 'IATI_VALIDATOR_QUOTA_PERIOD', timedelta(seconds=0))
```

Comment:
The zeros as default are intentional. The rate limit will not be applied when the rate has 0 in one of its attributes. This is useful, especially in tests to avoid excessive mocking.